### PR TITLE
Add batching function for activity replication task

### PR DESCRIPTION
**Pull request by xwduan**
## What changed?

Add batching functionality for activity replication task and gated with feature flag
## Why?

To improve the replication stack efficiency.
## How did you test it?

Unit test added
## Potential risks

Gated with feature flag. Risk is minimum.
## Is hotfix candidate?

No

### CODE REVIEW

**Comment by stephanos** [comment]
Is this ready for review? Asking since I don't see the What/Why filled out.

**Comment by xwduan** [comment]
> Is this ready for review? Asking since I don't see the What/Why filled out.

Was distracted when adding the details. Thank you for reminding.

File: service/history/replication/executable_activity_state_task.go [hunk 1]
```diff
@@ -209,3 +240,54 @@ func (e *ExecutableActivityStateTask) MarkPoisonPill() error {
...
 }
+
+func (e *ExecutableActivityStateTask) BatchWith(incomingTask BatchableTask) (TrackableExecutableTask, bool) {
+	if !e.batchable {
+		return nil, false
+	}
+	e.batchLock.Lock()
+	defer e.batchLock.Unlock()
+
+	if !e.batchable || !incomingTask.CanBatch() {
```
Comment (yux0) [inline service/history/replication/executable_activity_state_task.go:251]: move !incomingTask.CanBatch() before the lock?

**Review by yux0 — APPROVED** [review]

File: service/history/replication/executable_activity_state_task.go [hunk 2]
```diff
@@ -49,11 +50,17 @@ type (
 		definition.WorkflowKey
 		ExecutableTask
 		req *historyservice.SyncActivityRequest
+
+		// following fields are used only for batching functionality
+		batchLock     sync.Mutex
+		batchable     bool
```
Comment (tdeebswihart) [inline service/history/replication/executable_activity_state_task.go:56]: Mind separating this field from the mutex? I think we should only group mutexes with the fields they protect

File: service/history/replication/executable_activity_state_task.go [hunk 3]
```diff
@@ -209,3 +240,54 @@ func (e *ExecutableActivityStateTask) MarkPoisonPill() error {
 
 	return writeTaskToDLQ(ctx, e.DLQWriter, shardContext, e.SourceClusterName(), replicationTaskInfo)
 }
+
+func (e *ExecutableActivityStateTask) BatchWith(incomingTask BatchableTask) (TrackableExecutableTask, bool) {
+	if !e.batchable || !incomingTask.CanBatch() {
```
Comment (tdeebswihart) [inline service/history/replication/executable_activity_state_task.go:245]: You're accessing this outside of the lock here but only setting it inside the lock.

This is a race condition and could lead to undesireable behavior

File: service/history/replication/executable_activity_state_task.go [hunk 4]
```diff
@@ -209,3 +240,54 @@ func (e *ExecutableActivityStateTask) MarkPoisonPill() error {
...
+func (e *ExecutableActivityStateTask) CanBatch() bool {
+	e.batchLock.Lock()
+	defer e.batchLock.Unlock()
+	return e.batchable
+}
+
+func (e *ExecutableActivityStateTask) MarkUnbatchable() {
+	e.batchLock.Lock()
+	defer e.batchLock.Unlock()
+	e.batchable = false
```
Comment (tdeebswihart) [inline service/history/replication/executable_activity_state_task.go:292]: Batchable should only be accessed within the lock or outside of the lock as a sync/atomic.Bool or something. We're mixing access patterns here

File: service/history/replication/executable_activity_state_task.go (hunk 3 above)
Comment (xwduan) [reply service/history/replication/executable_activity_state_task.go:245]: Follow up PR to remove the lock: https://github.com/temporalio/temporal/pull/5280 please take a look when you got time :)

**Pull request merged**